net/tcp/tcp_conn.c

```diff
@@ -586,16 +586,32 @@ FAR struct tcp_conn_s *tcp_alloc_conn(void)
 
 void tcp_initialize(void)
 {
-#ifndef CONFIG_NET_ALLOC_CONNS
   int i;
 
+#ifndef CONFIG_NET_ALLOC_CONNS
   for (i = 0; i < CONFIG_NET_TCP_CONNS; i++)
     {
       /* Mark the connection closed and move it to the free list */
 
       g_tcp_connections[i].tcpstateflags = TCP_CLOSED;
       dq_addlast(&g_tcp_connections[i].sconn.node, &g_free_tcp_connections);
     }
+#else
+  FAR struct tcp_conn_s *conn;
+
+  for (i = 0; i < CONFIG_NET_TCP_PREALLOC_CONNS; i++)
+    {
+      conn = kmm_zalloc(sizeof(struct tcp_conn_s));
```

Comment:
why not reuse g_tcp_connections?

Reply:
When using `CONFIG_NET_ALLOC_CONNS` connections are dynamically allocated in the heap. I believe it is much better to have everything in the heap, than scattering things around.

It also keeps code simpler I think, and I don't see any drawbacks this way.

Reply:
> When using `CONFIG_NET_ALLOC_CONNS` connections are dynamically allocated in the heap. I believe it is much better to have everything in the heap, than scattering things around.

kmm_malloc has the 8bytes overhead for each allocation. If you will always allocate something initially, it's better to either define a global array or allocate in a batch.

> 
> It also keeps code simpler I think,

But not, now we have two code path to initialize conn. Even worse, we have two option to CONFIG_NET_TCP_PREALLOC_CONNS and CONFIG_NET_TCP_ALLOC_CONNS for the similar config.

> and I don't see any drawbacks this way.

Reply:
If you really  want to allocate and free conn dynamically, it could be achieved simply by:
Set CONFIG_NET_CONNS to 1 and kmm_free conn in case CONFIG_NET_CONNS equals 1 instead put it into free list.

Reply:
The reason that I do not want to use `g_tcp_connections` is because this way we will have connections allocated both in heap and BSS,
and make memory management much more complicated (linker scripts etc). Especially in arches with multiple memory regions.

It is much better to have everything in one place. Custom memory mappings and adjustment of the settings are becoming much easier.

You are right about the heap overhead. I will change this to allocate all connections in a single `kmm_zalloc()` call.

> But not, now we have two code path to initialize conn.

This is a very simple change, all contained within the same function. It is not really complicating things.

> Even worse, we have two option to CONFIG_NET_TCP_PREALLOC_CONNS and NET_TCP_CONNS for the similar config.

I did this for clarity. But it is no problem to always use `NET_TCP_CONNS`. I will do this change soon.

> If you really want to allocate and free conn dynamically, it could be achieved simply by:
Set CONFIG_NET_CONNS to 1 and kmm_free conn in case CONFIG_NET_CONNS equals 1 instead put it into free list.

As I explained before, this is a different thing. With this approach we completely lose the benefit of having preallocated connections.

Reply:
> The reason that I do not want to use `g_tcp_connections` is because this way we will have connections allocated both in heap and BSS, and make memory management much more complicated (linker scripts etc). Especially in arches with multiple memory regions.
> 
> It is much better to have everything in one place. Custom memory mappings and adjustment of the settings are becoming much easier.
> 

If so, you can add a new option(e.g. NET_TCP_PREALLOC_CONNS to indicate the preallocated global array, but keep NET_TCP_ALLOC_CONNS as the dynamical allocation:

1. Set NET_TCP_PREALLOC_CONNS to zero to disable the global allocation
2. Set NET_TCP_ALLOC_CONNS to 1 to free conn instead put into a free list

Reply:
> I sincerely do not understand your proposition. I need the following:
> 
> * It is impossible to know the maximum number of connections before-hand. Thus I need dynamic connections allocation.
> * I know that a temporal high-load condition may happen. After this finishes, the connections must be deallocated to allow the system to reuse the memory. It must not keep the never-to-be-used-again memory reserved!

This can simply achieve to set CONFIG_NET_CONNS to 1, and alloc/free conn.

> * I know that at least _some_ connections are always needed, and I need them preallocated to reduce heap fragmentation, enhance performance, and reduce locking of the system (as the memory manager may do).
> 

it can achieve with new NET_TCP_PREALLOC_CONNS, you can allocate from .bss or heap in batch, both approach is fine.

> All the above sound like very reasonable and possibly common requirements for a networking application. In fact, I cannot imagine a system with unknown number of connections behave any differently.

Yes, I think all is the reasonable requirement.

> Point 2 is especially important for safety and security reasons.
> 

Point 2 is important for the device which accept the connection from client, but many IoT device only contain client side functionality and never accept the new connection. In the second case, it's safe to not release conn to heap, but put into the free list:

1. Allocate conn in batch to avoid the heap overhead
2. Avoid the memory fragment by reducing malloc/free operation
3. Improve the speed

> Your proposal cannot satisfy all 3 requirements, at least as far as I can understand it.

Please reference my suggestion here:
https://github.com/apache/nuttx/pull/7525#discussion_r1038789173
And point out which's function you want can't achieve, let improve it case by case.